### PR TITLE
[WIP]: Add deferrable mode to `RedshiftDataOperator`

### PR DESCRIPTION
**Pull request by phanikumv**
This PR adds the `RedshiftDataOperatorAsync` from [astronomer-providers ](https://github.com/astronomer/astronomer-providers) to Airflow

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 1]
```diff
@@ -190,3 +198,198 @@ def get_table_primary_key(
...
+            if isinstance(sql, str):
+                split_statements_tuple = split_statements(StringIO(sql))
+                sql = [sql_string for sql_string, _ in split_statements_tuple if sql_string]
+
+            try:
+                # for apache-airflow-providers-amazon>=3.0.0
+                client = self.get_conn()
+            except ValueError:
+                # for apache-airflow-providers-amazon>=4.1.0
+                self.resource_type = None
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:320]: ???

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 2]
```diff
@@ -190,3 +198,198 @@ def get_table_primary_key(
...
+        (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html)
+    :param poll_interval: polling period in seconds to check for the status
+    """
+
+    def __init__(self, *args: Any, poll_interval: int = 0, **kwargs: Any) -> None:
+        aws_connection_type: str = "redshift-data"
+        try:
+            # for apache-airflow-providers-amazon>=3.0.0
+            kwargs["client_type"] = aws_connection_type
+            kwargs["resource_type"] = aws_connection_type
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:229]: This is not valid to set both `client_type` and `resource_type` and also `aibotocore` do not support resources

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 3]
```diff
@@ -190,3 +198,198 @@ def get_table_primary_key(
...
+    :param region_name: AWS region_name. If not specified then the default boto3 behaviour is used.
+    :param client_type: boto3.client client_type. Eg 's3', 'emr' etc
+    :param resource_type: boto3.resource resource_type. Eg 'dynamodb' etc
+    :param config: Configuration for botocore client.
+        (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html)
+    :param poll_interval: polling period in seconds to check for the status
+    """
+
+    def __init__(self, *args: Any, poll_interval: int = 0, **kwargs: Any) -> None:
+        aws_connection_type: str = "redshift-data"
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:225]: This connection not provided by Amazon Provider

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 4]
```diff
@@ -17,11 +17,19 @@
...
+import asyncio
+from io import StringIO
 from pprint import pformat
 from time import sleep
 from typing import TYPE_CHECKING, Any, Iterable
 
-from airflow.providers.amazon.aws.hooks.base_aws import AwsGenericHook
+import botocore.exceptions
+from asgiref.sync import sync_to_async
+from snowflake.connector.util_text import split_statements
```
Comment (eladkal) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:28]: snowflake? why?

File: airflow/providers/amazon/aws/hooks/redshift_data.py [hunk 5]
```diff
@@ -190,3 +198,198 @@ def get_table_primary_key(
...
+            # for apache-airflow-providers-amazon>=4.1.0
+            kwargs["client_type"] = aws_connection_type
+            super().__init__(*args, **kwargs)
+        self.client_type = aws_connection_type
+        self.poll_interval = poll_interval
+
+    def get_conn_params(self) -> dict[str, str | int]:
+        """Helper method to retrieve connection args"""
+        if not self.aws_conn_id:
+            raise AirflowException("Required connection details is missing !")
```
Comment (Taragolis) [inline airflow/providers/amazon/aws/hooks/redshift_data.py:241]: - https://airflow.apache.org/docs/apache-airflow-providers-amazon/7.3.0/connections/aws.html#per-service-configuration
- https://github.com/apache/airflow/pull/28706#discussion_r1062824388

**Pull request closed without merging**